### DISCUSSION TITLE
Will demo-portal be ported to 3.0 any time soon?

### DISCUSSION THREAD

**Discussion by JohnMajerle** [iTwin/itwinjs-core#3058] [Q&A] [answered]
My customers have showed considerable interest in the UI features from a prototype I created based on **_demo-portal:_**

     https://github.com/iTwin/demo-portal

What is Bentley's timeline for updating this git repo to iTwin 3.0, or even better, deliver a starter templated based on a 3.0 update of this git repo?

If there are no plans, how effective will running **_codemon_** be converting this code given that iTwin 3.0 RC still has some features that are not yet ported?

And finally, does Bentley have a target date for officially releasing iTwin 3.0?

Thanks

**Comment by calebmshafer** [comment]
Hey @JohnMajerle, we are getting ready to release iTwin.js 3.0 this week. There will be a post shortly with the list of items being updated.

For a template are you referring to an updated iTwin Viewer template or for the demo portal itself?

cc: @Josh-Schifter for updating the demo portal. I don't know the status of this update.

**Comment by JohnMajerle** [comment]
Hello Caleb,
I'm talking about the demo portal.  Our customer wasn't impressed with the iTwin Viewer so it's a non-starter with us.  They much preferred the user experience from a prototype we created for them based on demo portal.  And for obvious reasons, going forward we'd prefer to deliver an application based on the current code base instead of building on top of the old code base so that we can take advantage of all the newer iTwin API features and bug fixes Bentley is providing.

**Reply by calebmshafer** [reply]
okay, sounds good @JohnMajerle. The iTwin Viewer is one piece of the Demo Portal in the "View iModel" tab but I'm guessing you mean the overall application layout is a more desirable starting point for your application then the iTwin Viewer template. Did I get that right?

We will try and run the codemod tool on the existing demo portal and see what gaps there are. Did you have any specific examples when referring to the tool with "still has some features that are not yet ported"? If so, we can look at tackling those asap so you could run it on your prototype.

**Reply by JohnMajerle** [reply]
Hello Caleb ( @calebmshafer ),
Did you have any luck porting **demo portal** to 3.0?  I made some progress using codemod on my source but am running into numerous API changes that are giving me grief.  It would be a big help if I could start with a clean 3.0 **demo portal** and then incrementally graft on my business logic.

**Comment by JohnMajerle** [comment]
Hello Caleb ( @calebmshafer ),
Yes, the demo portal overall layout is much better.  Among other things, the demo portal provides an intuitive UI for iModel selection.   It includes a variety of useful UI related to project and iModel management.  And overall is just looks more professional.
The feature I was concerned about that was mentioned in the RC3 announcement was the missing element properties.  Beyond that I don't know if anything else is missing but that's a pretty important feature.  Our application emphasizes quality assurance of civil engineering data so it isn't very helpful for us to identify problematic elements if the user can't then review these element properties in the UI. 
There are a variety of other UI enhancements I'd like to request but I will include those in different discussion(s) once we can move to 3.0.

**Reply by aruniverse** [reply]
Hi John,

Regarding the missing features in the initial RC announcement, please see https://github.com/iTwin/itwinjs-core/discussions/2942#discussioncomment-2000206. The tree-widget and property-grid have been added back into the viewer.

Regarding the features of the demo-portal like imodel selection, those should all still be available to use. Check out the admin-components-react repo, https://github.com/iTwin/admin-components-react/tree/main/packages/modules. These packages should be decoupled from itwinjs-core (since all they do is just make some REST reqs) and dont need to be updated to 3.0. Using the codemod should mostly work for you, you might just have to react to some breaking changes.

**Comment by JohnMajerle** [comment]
Hello Arun ( @aruniverse )

I made the change to my package.json and reexecuted both **_yarn_** and **_yarn start._**

I am getting the following runtime error:

Failed to compile.

./node_modules/@itwin/appui-react/lib/appui-react/pickers/ListPicker.scss (./node_modules/css-loader/dist/cjs.js??ref--7-oneOf-5-1!./node_modules/postcss-loader/src??postcss!./node_modules/resolve-url-loader??ref--7-oneOf-5-3!./node_modules/sass-loader/dist/cjs.js??ref--7-oneOf-5-4!./node_modules/@itwin/appui-react/lib/appui-react/pickers/ListPicker.scss)
SassError: Can't find stylesheet to import.
  ╷
5 │ @import "~@itwin/appui-layout-react/lib/appui-layout-react/footer/_variables";
  │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
  Z:\iTwin30RC\datamanager-portal\packages\app\node_modules\@itwin\appui-react\lib\appui-react\pickers\ListPicker.scss 5:9  root stylesheet

**Reply by aruniverse** [reply]
Hi John, yes that's expected. Please see the following breaking changes wrt scss, https://github.com/iTwin/itwinjs-core/blob/release/3.0.x/docs/changehistory/NextVersion.md#buildology

**Comment by JohnMajerle** [comment]
Hello Arun ( @aruniverse ),

I don't understand.  This error is happening in an itwin package, not in my code.  What do I need to change to resolve it?

**Reply by aruniverse** [reply]
Sorry John, can you take a look at your yarn.lock lockfile, and reply back with the version of itwin/appui-layout-react

**Comment by JohnMajerle** [comment]
Hello Arun (  @aruniverse  )

"@itwin/appui-layout-react@rc":
  version "3.0.0-dev.185"
  resolved "https://registry.yarnpkg.com/@itwin/appui-layout-react/-/appui-layout-react-3.0.0-dev.185.tgz#b07dd3a5bcfb411b5e54231fecdf65bdb68c8ee7"
  integrity sha512-58qlaE0XuhdC9LVQnFM8wimY0nVkpkoReGfHNi1CI+dafLVzVQGStdmqWA6B+si8ncXspzduQaNUbXJdxW08Eg==
  dependencies:
    "@itwin/itwinui-css" "^0.37.0"
    "@itwin/itwinui-react" "~1.28.0"
    classnames "^2.3.1"
    immer "9.0.6"
    uuid "^7.0.3"

**Reply by JohnMajerle** [reply]
Hello Arun ( @aruniverse ),
Was that information helpful?  
Thanks,

**Reply by aruniverse** [reply]
Ah sorry I didnt get back to you John. The version looks right, I was grepping through our src and didnt see where that offending line could be coming from in the core packages. Mind nuking all your node_modules, and installing again? If the error still occurs, we need to find whats pulling in that scss file

**Comment by JohnMajerle** [comment]
Hello Arun  ( @aruniverse ),

I deleted the 3 **node_modules** directories and reinstalled with yarn.  Yarn start gets the same error.

Below is the contents of my: Z:\iTwin30RC\datamanager-portal\packages\app\package.json.

From looking at this I don't understand the @itwin versioning standard.  Should all of these be set to "rc" instead of a particular dev version?  Or should all of them be set to "^3.0.0-dev.72" ?

Thanks,

{
  "name": "itwin-demo-portal",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@itwin/core-bentley": "^3.0.0-dev.72",
    "@bentley/extension-client": "^2.19.3",
    "@bentley/frontend-application-insights-client": "^2.19.3",
    "@itwin/browser-authorization": "latest",
    "@itwin/core-geometry": "rc",
    "@bentley/icons-generic": "^1.0.13",
    "@bentley/imodelhub-client": "^2.19.3",
    "@itwin/core-common": "rc",
    "@itwin/core-frontend": "rc",
    "@itwin/core-i18n": "rc",
    "@itwin/core-markup": "rc",
    "@itwin/core-quantity": "rc",
    "@bentley/itwin-client": "^2.19.3",
    "@itwin/core-orbitgt": "rc",
    "@itwin/presentation-common": "rc",
    "@itwin/presentation-frontend": "rc",
    "@bentley/product-settings-client": "^2.19.3",
    "@bentley/rbac-client": "^2.19.3",
    "@bentley/telemetry-client": "^2.19.3",
    "@itwin/appui-abstract": "^3.0.0-dev.72",
    "@itwin/components-react": "rc",
    "@itwin/core-react": "rc",
    "@itwin/appui-react": "^3.0.0-dev.72",
    "@itwin/appui-layout-react": "rc",
    "@itwin/webgl-compatibility": "rc",
    "@itwin/create-imodel-react": "^0.4.1",
    "@itwin/delete-imodel-react": "^0.2.0",
    "@itwin/imodel-browser-react": "^0.11.0",
    "@itwin/itwinui-css": "^0.18.1",
    "@itwin/itwinui-icons-react": "^1.2.0",
    "@itwin/itwinui-react": "^1.15.0",
    "@itwin/manage-versions-react": "^0.7.0",
    "@itwin/web-viewer-react": "^2.0.0-dev.21",
    "@microsoft/applicationinsights-react-js": "^3.1.3",
    "@microsoft/applicationinsights-web": "^2.6.4",
    "@reach/router": "^1.3.4",
    "@types/react-datepicker": "^4.3.2",
    "byte-size": "^8.1.0",
    "history": "^4.10.1",
    "isomorphic-fetch": "^3.0.0",
    "launchdarkly-js-sdk-common": "^3.3.2",
    "launchdarkly-react-client-sdk": "^2.22.2",
    "oidc-client": "^1.10.1",
    "react": "16.13.1",
    "react-datepicker": "^4.3.0",
    "react-dom": "16.13.1",
    "react-intersection-observer": "^8.32.0",
    "styled-components": "^5.3.3"
  },
  "devDependencies": {
    "@bentley/react-scripts": "^3.4.1",
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/react": "^10.4.4",
    "@testing-library/user-event": "^7.1.2",
    "@types/history": "^4.7.6",
    "@types/jest": "^26.0.4",
    "@types/node": "^14.0.19",
    "@types/reach__router": "^1.3.7",
    "@types/react": "^16.9.0",
    "@types/react-dom": "^16.9.0",
    "msw": "~0.32.0",
    "typescript": "^3.7.2"
  },
  "scripts": {
    "start": "react-scripts --max_old_space_size=4096 start",
    "build": "react-scripts --max_old_space_size=4096 build",
    "test": "react-scripts --max_old_space_size=4096 test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "../../.eslintrc.ts.json"
    ]
  },
  "browserslist": [
    "electron 8.0.0",
    "last 4 chrome version",
    "last 4 firefox version",
    "last 4 safari version",
    "last 4 ios version",
    "last 4 ChromeAndroid version",
    "last 4 edge version",
    "not dead",
    "not <0.2%"
  ],
  "resolutions": {
    "faye-websocket": "^0.11.0",
    "ua-parser-js": "^0.7.22",
    "dompurify": "^2.1.0",
    "object-path": "^0.11.5",
    "axios": "0.21.2",
    "immer": "^9.0.6"
  }
}

**Reply by aruniverse** [reply]
For 3.0, we are moving our core packages (from the itwinjs-core repo) to be published under the itwin scope, see https://github.com/iTwin/itwinjs-core/blob/release/3.0.x/docs/changehistory/NextVersion.md#package-name-changes.

Any core package that you're using with the bentley scope has either been renamed to one under the itwin scope or eliminated altogether. Just looking at the list above, the following needs to be removed (as they're incompatible with 3.0):
- @bentley/extension-client
- @bentley/frontend-application-insights-client
- @bentley/imodelhub-client
- @bentley/product-settings-client
- @bentley/rbac-client
- @bentley/telemetry-client

To test out the rc, I would recommend you don't set the version for the core pkgs to be "^3.0.0-dev.72" but rather use the "rc" dist tag. Hopefully, this should help the dependencies in your application be consistent and compatible.

**Comment by aruniverse** [comment, marked as answer]
The demo-portal has been updated to use 3.x; thanks @raplemie
